include/deal.II/multigrid/mg_transfer_global_coarsening.h

```diff
@@ -0,0 +1,444 @@
+  const MGLevelObject<MGTwoLevelTransfer<dim, VectorType>> &transfer;
+};
+
+
+
+#ifndef DOXYGEN
+
+/* ----------------------- Inline functions --------------------------------- */
+
+
+
+template <int dim, typename Number>
+template <typename Stream>
+void
+MGTwoLevelTransfer<dim, LinearAlgebra::distributed::Vector<Number>>::
+  MGTransferScheme::print(Stream &out) const
+{
+  out << "weights:" << std::endl;
+  for (const auto w : weights)
+    out << w << " ";
+  out << std::endl;
+
+  out << "level_dof_indices_fine:" << std::endl;
+  for (const auto w : level_dof_indices_fine)
+    out << w << " ";
+  out << std::endl;
+
+  out << "level_dof_indices_coarse:" << std::endl;
+  for (const auto w : level_dof_indices_coarse)
+    out << w << " ";
+  out << std::endl;
+
+  out << "prolongation_matrix_1d:" << std::endl;
+  for (const auto w : prolongation_matrix_1d)
+    out << w[0] << " ";
+  out << std::endl;
+}
+
+
+
+template <int dim, typename Number>
+template <typename Stream>
+void
+MGTwoLevelTransfer<dim, LinearAlgebra::distributed::Vector<Number>>::
+  print_internal(Stream &out) const
+{
+  for (const auto &scheme : schemes)
+    scheme.print(out);
+}
+
+
+
+template <typename MatrixType, typename VectorType>
+MGTransferGlobalCoarsening<MatrixType, VectorType>::MGTransferGlobalCoarsening(
+  const MGLevelObject<MatrixType> &                         matrices,
+  const MGLevelObject<MGTwoLevelTransfer<dim, VectorType>> &transfer)
+  : matrices(matrices)
+  , transfer(transfer)
+{
+  AssertDimension(matrices.max_level() - matrices.min_level(),
+                  transfer.max_level() - transfer.min_level());
+}
+
+
+
+template <typename MatrixType, typename VectorType>
+void
+MGTransferGlobalCoarsening<MatrixType, VectorType>::prolongate(
+  const unsigned int to_level,
+  VectorType &       dst,
+  const VectorType & src) const
+{
+  this->transfer[to_level].prolongate(numbers::invalid_unsigned_int /*dummy*/,
+                                      dst,
+                                      src);
+}
+
+
+
+template <typename MatrixType, typename VectorType>
+void
+MGTransferGlobalCoarsening<MatrixType, VectorType>::restrict_and_add(
+  const unsigned int from_level,
+  VectorType &       dst,
+  const VectorType & src) const
+{
+  this->transfer[from_level].restrict_and_add(
+    numbers::invalid_unsigned_int /*dummy*/, dst, src);
+}
+
+
+
+template <typename MatrixType, typename VectorType>
+template <class InVector, int spacedim>
+void
+MGTransferGlobalCoarsening<MatrixType, VectorType>::copy_to_mg(
+  const DoFHandler<dim, spacedim> &dof_handler,
+  MGLevelObject<VectorType> &      dst,
+  const InVector &                 src) const
+{
+  (void)dof_handler;
+
+  for (unsigned int level = dst.min_level(); level <= dst.max_level(); ++level)
+    matrices[level].initialize_dof_vector(dst[level]);
+
+  dst[dst.max_level()].copy_locally_owned_data_from(src);
+  dst[dst.max_level()].update_ghost_values();
```

Comment:
Why do you need `update_ghost_values` here? To me, ghost values should only be held close to the place where we need to read remote vector entries; otherwise all global vector operations like `Vector::add()` call `update_ghost_values` which is likely not intended.